[Merged by Bors] - feat(Algebra/Order/Ring/WithTop): add generalised and specialised power of top

pitmonticone opened the pull request:
- [x] Add generalised version of `top_pow` in `Algebra/Order/Ring/WithTop`
- [x] Add specialised version of `top_pow` in `Data/ENat/Basic.lean`
- [x] Add specialised version of `top_pow` in `Data/ENNReal/Operations.lean` 

Co-authored by @D-Thomine.

sgouezel commented on Mathlib/Data/ENat/Basic.lean:
```diff
@@ -89,6 +89,16 @@ theorem coe_sub (m n : ℕ) : ↑(m - n) = (m - n : ℕ∞) :=
 @[simp] theorem top_mul (hm : m ≠ 0) : ⊤ * m = ⊤ := WithTop.top_mul hm
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℕ∞)^n = ⊤ := by
+  apply @Nat.le_induction 1 (fun m : ℕ ↦ fun _ : 1 ≤ m ↦ (⊤ : ℕ∞) ^ m = ⊤) (pow_one ⊤)
+  · intro m _ h
+    calc
+      (⊤ : ℕ∞)^(m + 1) = ⊤^m * ⊤^1 := by rw [pow_add ⊤ m 1]
+                     _ = ⊤ * ⊤^1   := by rw [h]
+                     _ = ⊤ * ⊤     := by rw [pow_one ⊤]
+                     _ = ⊤         := WithTop.top_mul_top
+  · exact n_pos
```
When the steps of the computation are trivial, it's often better to just use `simp`, like
```suggestion
  apply Nat.le_induction (P := fun m : ℕ ↦ fun _ : 1 ≤ m ↦ (⊤ : ℕ∞) ^ m = ⊤) (pow_one ⊤) _ _ n_pos
  intro m _ h
  simp only [pow_add, h, pow_one]
  exact WithTop.top_mul_top
```

pitmonticone commented:
Ok, I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

sgouezel commented:
In fact, this should be generalized to `WithTop`. Could you take the proof from `ENNReal.top_pow` and move it to `Algebra.Order.Ring.WithTop` at Line 170 (and then restate `ENNReal.top_pow` and `ENat.top_pow` as specialized versions of the general statement?)

pitmonticone commented:
I should have addressed all the review requests: 

- [x] Add generalised version of `top_pow` in `Algebra/Order/Ring/WithTop`
- [x] Add specialised version of `top_pow` in `Data/ENat/Basic.lean`
- [x] Add specialised version of `top_pow` in `Data/ENNReal/Operations.lean` 

Thank you @sgouezel for your review.

sgouezel commented:
My comment was to copy the proof of `ENNReal.top_pow` because it is better than the proof of the current PR. The comment still stands :-)

sgouezel commented on Mathlib/Algebra/Order/Ring/WithTop.lean:
```diff
@@ -167,6 +167,14 @@ instance instMonoidWithZero : MonoidWithZero (WithTop α) where
+  intro m _ h
+  simp only [pow_add, h, pow_one]
+  exact top_mul_top
+#align withtop.top_pow WithTop.top_pow
```
no #align for new statements. It's only useful for statements that were already existing in Lean 3, for porting tools.

sgouezel commented on Mathlib/Data/ENat/Basic.lean:
```diff
@@ -89,6 +89,9 @@ theorem coe_sub (m n : ℕ) : ↑(m - n) = (m - n : ℕ∞) :=
 @[simp] theorem top_mul (hm : m ≠ 0) : ⊤ * m = ⊤ := WithTop.top_mul hm
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℕ∞)^n = ⊤ := WithTop.top_pow n_pos
+#align enat.top_pow ENat.top_pow
```
ditto

sgouezel replied:
The "no #align for new statements" also applies here. That's the meaning of "ditto".

pitmonticone replied:
No, I did not... sorry. 

Done now.

sgouezel commented on Mathlib/Algebra/Order/Ring/WithTop.lean:
```diff
@@ -167,6 +167,14 @@ instance instMonoidWithZero : MonoidWithZero (WithTop α) where
 
 @[simp, norm_cast] lemma coe_pow (a : α) (n : ℕ) : (↑(a ^ n) : WithTop α) = a ^ n := rfl
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : WithTop α)^n = ⊤ := by
```
```suggestion
theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : WithTop α) ^ n = ⊤ := by
```
Spaces in power operation, here and elsewhere in the PR.

sgouezel commented on Mathlib/Data/ENNReal/Operations.lean:
```diff
@@ -219,10 +219,7 @@ theorem top_mul_top : ∞ * ∞ = ∞ := WithTop.top_mul_top
-theorem top_pow {n : ℕ} (h : 0 < n) : ∞ ^ n = ∞ :=
-  Nat.le_induction (pow_one _) (fun m _ hm => by rw [pow_succ, hm, top_mul_top]) _
-    (Nat.succ_le_of_lt h)
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℝ≥0∞)^n = ⊤ := WithTop.top_pow n_pos
```
In ENNReal, the notation ∞ is preferred.

pitmonticone commented:
Ok, I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

pitmonticone commented:
Now I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

sgouezel commented:
bors r+
Thanks!
